network/builder/src/builder.rs

```diff
@@ -183,6 +183,99 @@ impl NetworkBuilder {
+                role,
+                peer_id,
+                config.listen_address.clone(),
+            );
+            network_builder.add_connection_monitoring();
+
+            // Sanity check seed peer addresses.
+            config
+                .verify_seed_peer_addrs()
+                .expect("Seed peer addresses must be well-formed");
+            let seed_peers = config.seed_peers.clone();
+
+            if config.mutual_authentication {
+                let network_peers = config.network_peers.clone();
+                let trusted_peers = if role == RoleType::Validator {
+                    // for validators, trusted_peers is empty will be populated from consensus
+                    HashMap::new()
+                } else {
+                    network_peers
+                };
+
+                info!(
+                    "network setup: role: {}, seed_peers: {:?}, trusted_peers: {:?}",
+                    role, seed_peers, trusted_peers,
+                );
+
+                network_builder
+                    .authentication_mode(AuthenticationMode::Mutual(identity_key))
+                    .trusted_peers(trusted_peers)
+                    .seed_peers(seed_peers)
+                    .connectivity_check_interval_ms(config.connectivity_check_interval_ms)
+                    .add_connectivity_manager();
+            } else {
+                // Enforce the outgoing connection (dialer) verifies the identity of the listener (server)
+                network_builder.authentication_mode(AuthenticationMode::ServerOnly(identity_key));
+                if !seed_peers.is_empty() {
+                    network_builder
+                        .seed_peers(seed_peers)
+                        .add_connectivity_manager();
+                }
+            }
+
+            match &config.discovery_method {
+                DiscoveryMethod::Gossip(gossip_config) => {
+                    network_builder
+                        .advertised_address(gossip_config.advertised_address.clone())
+                        .discovery_interval_ms(gossip_config.discovery_interval_ms)
+                        .add_gossip_discovery();
+                }
+                DiscoveryMethod::Onchain => {
+                    let (network_tx, discovery_events) = network_builder.add_protocol_handler(
+                        onchain_discovery::network_interface::network_endpoint_config(),
+                    );
+                    let onchain_discovery_builder = OnchainDiscoveryBuilder::build(
+                        network_builder
+                            .conn_mgr_reqs_tx()
+                            .expect("ConnectivityManager must be installed"),
+                        network_tx,
+                        discovery_events,
+                        network_builder.network_context().clone(),
+                        libra_db,
+                        waypoint,
+                        runtime.handle(),
+                    );
```

Comment:
in a future PR, let's make this the simple on chain discovery... and if it is set and / or seed_peers is set and we're not in mutual_authentication, we want to add connectivity manager.

Reply:
An implied corollary is removing onchain discovery?

Reply:
Removing, not quite, we likely need to postpone to after V1. I imagine it will be quite useful to our ecosystem for light clients -- which we currently have yet to implement :).